src/engine/sparqlExpressions/BlankNodeExpression.cpp

```diff
@@ -0,0 +1,167 @@
+          ++counter_;
+        },
+        [context]() { context->cancellationHandle_->throwIfCancelled(); });
+    return result;
+  }
+
+  // __________________________________________________________________________
+  ExpressionResult evaluate(EvaluationContext* context) const override {
+    if (std::holds_alternative<size_t>(label_)) {
+      return evaluateImpl(context, [this]() -> std::optional<size_t> {
+        return std::get<size_t>(label_);
+      });
+    }
+    auto childResult = std::get<Ptr>(label_)->evaluate(context);
+    return std::visit(
+        [this, context]<typename T>(T&& element) -> ExpressionResult {
+          if constexpr (isConstantResult<T>) {
+            const auto& value = StringValueGetter{}(AD_FWD(element), context);
+            if (!value.has_value()) {
+              // Increment the counter for every element in the context for
+              // consistency.
+              counter_ += context->size();
+              return Id::makeUndefined();
+            }
+            auto escapedValue = escapeStringForBlankNode(value.value());
+            return evaluateImpl(
+                context, [&escapedValue]() -> std::optional<std::string_view> {
+                  return escapedValue;
+                });
+          } else {
+            auto generator = detail::makeGenerator(AD_FWD(element),
+                                                   context->size(), context);
+            return evaluateImpl(
+                context,
+                [it = generator.begin(), &generator,
+                 context]() mutable -> std::optional<std::string> {
+                  AD_CORRECTNESS_CHECK(it != generator.end());
+                  const auto& value =
+                      StringValueGetter{}(std::move(*it), context);
+                  ++it;
+                  if (!value.has_value()) {
+                    return std::nullopt;
+                  }
+                  return escapeStringForBlankNode(value.value());
+                });
+          }
+        },
+        std::move(childResult));
+  }
+
+  // ___________________________________________________________________________
+  std::string getCacheKey(const VariableToColumnMap& map) const override {
+    if (std::holds_alternative<size_t>(label_)) {
+      return absl::StrCat("#BlankNode#", std::get<size_t>(label_), "_",
+                          cacheBreaker_++);
+    }
+    return absl::StrCat("#BlankNode#", std::get<Ptr>(label_)->getCacheKey(map),
+                        "_", cacheBreaker_++);
+  }
+
+ private:
+  std::span<Ptr> childrenImpl() override {
+    if (std::holds_alternative<Ptr>(label_)) {
+      return {&std::get<Ptr>(label_), 1};
+    }
+    return {};
+  }
+};
+}  // namespace detail
+
+SparqlExpression::Ptr makeBlankNodeExpression(SparqlExpression::Ptr child) {
+  return std::make_unique<detail::BlankNodeExpression>(std::move(child));
+}
+SparqlExpression::Ptr makeUniqueBlankNodeExpression(size_t label) {
```

Comment:
The `label` based approach is rather hacky.
In particular it breaks with persisted delta triples on Updates, 
which is somewhat bad. 
I will try to make this work somewhat cleaner.